File: app/src/main/java/com/iseasoft/iseaiptv/ui/activity/MainActivity.java
package com.iseasoft.iseaiptv.ui.activity;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.support.constraint.ConstraintLayout;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.NavigationView;
import android.support.design.widget.Snackbar;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentStatePagerAdapter;
import android.support.v4.view.GravityCompat;
import android.support.v4.view.ViewPager;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.iseasoft.iseaiptv.Constants;
import com.iseasoft.iseaiptv.R;
import com.iseasoft.iseaiptv.helpers.Router;
import com.iseasoft.iseaiptv.http.HttpHandler;
import com.iseasoft.iseaiptv.models.M3UItem;
import com.iseasoft.iseaiptv.models.M3UPlaylist;
import com.iseasoft.iseaiptv.models.Playlist;
import com.iseasoft.iseaiptv.parsers.M3UParser;
import com.iseasoft.iseaiptv.permissions.IseaSoft;
import com.iseasoft.iseaiptv.permissions.PermissionCallback;
import com.iseasoft.iseaiptv.ui.fragment.ChannelFragment;
import com.iseasoft.iseaiptv.utils.PreferencesUtility;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

public class MainActivity extends BaseActivity
        implements NavigationView.OnNavigationItemSelectedListener {

    private int allChannelTabIndex = 1;
    private CoordinatorLayout panelLayout;
    private ViewPager viewPager;
    private TabLayout tabLayout;
    private ConstraintLayout placeholderContainer;
    private ProgressBar progressBar;

    private M3UPlaylist mPlaylist;
    private GroupChannelAdapter adapter;
    private final PermissionCallback permissionReadstorageCallback = new PermissionCallback() {
        @Override
        public void permissionGranted() {
            loadChannels();
        }

        @Override
        public void permissionRefused() {
            requestStoragePermission();
        }
    };

    public M3UPlaylist getPlaylist() {
        return mPlaylist;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        setContentView(R.layout.activity_main);
        super.onCreate(savedInstanceState);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(
                this, drawer, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        drawer.addDrawerListener(toggle);
        toggle.syncState();

        NavigationView navigationView = (NavigationView) findViewById(R.id.nav_view);
        navigationView.setNavigationItemSelectedListener(this);

        panelLayout = findViewById(R.id.panel_layout);
        viewPager = (ViewPager) findViewById(R.id.viewpager);
        tabLayout = (TabLayout) findViewById(R.id.tabs);
        tabLayout.setupWithViewPager(viewPager);
        placeholderContainer = findViewById(R.id.placeholder_container);
        progressBar = findViewById(R.id.progressBar);

        checkToPlayFromPushNotification();
        loadChannels();
    }

    private void checkToPlayFromPushNotification() {
        Intent intent = getIntent();
        if (intent != null) {
            String url = intent.getStringExtra(Constants.PUSH_URL_KEY);
            String message = intent.getStringExtra(Constants.PUSH_MESSAGE);

            if (!TextUtils.isEmpty(url)) {
                if (checkPlayableUrl(url)) {
                    Intent playerIntent = new Intent(this, PlayerActivity.class);
                    playerIntent.putExtra(Constants.PUSH_URL_KEY, url);
                    playerIntent.putExtra(Constants.PUSH_MESSAGE, message);
                    startActivity(playerIntent);
                } else {
                    Playlist playlist = new Playlist();
                    playlist.setName(message);
                    playlist.setLink(url);
                    PreferencesUtility.getInstance(this).savePlaylist(playlist);
                }
            }
        }
    }

    private boolean checkPlayableUrl(String url) {
        if (url.contains(".m3u8") || url.contains(".ts") || url.contains(".mp4")) {
            return true;
        }
        return false;
    }

    @Override
    protected void onResume() {
        super.onResume();
    }

    private void loadChannels() {
        final Playlist lastPlaylist = PreferencesUtility.getInstance(this).getLastPlaylist();
        if (lastPlaylist != null) {
            displayPlaylistInfo(lastPlaylist);
            if (lastPlaylist.getLink().startsWith("http")) {
                loadServer(lastPlaylist.getLink());
            } else {
                try {
                    File file = new File(lastPlaylist.getLink());
                    InputStream inputStream = new FileInputStream(file);
                    parsePlaylist(inputStream);
                } catch (FileNotFoundException e) {
                    e.printStackTrace();
                    updateUI();
                }
            }
        } else {
            updateUI();
        }

    }

    private void displayPlaylistInfo(Playlist lastPlaylist) {
        NavigationView navigationView = (NavigationView) findViewById(R.id.nav_view);
        View header = navigationView.getHeaderView(0);
        TextView playlistName = header.findViewById(R.id.nav_header_title);
        TextView playlistLink = header.findViewById(R.id.nav_header_description);
        playlistName.setText(lastPlaylist.getName());
        //playlistLink.setText(lastPlaylist.getLink());
        getSupportActionBar().setTitle(lastPlaylist.getName());
    }

    private void showChannelPlaceholder() {
        tabLayout.setVisibility(View.GONE);
        progressBar.setVisibility(View.GONE);
        placeholderContainer.setVisibility(View.VISIBLE);
        Button btnAdd = findViewById(R.id.btn_add_playlist);
        btnAdd.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                navigateToPlaylist();
            }
        });
    }

    private void setupViewPager(ViewPager viewPager) {
        if (adapter == null) {
            adapter = new GroupChannelAdapter(getSupportFragmentManager());
        }
        adapter.addFragment(getString(R.string.favorites));
        if (!PreferencesUtility.getInstance(this).hasNoHistoryWatching()) {
            adapter.addFragment(getString(R.string.history_watching));
            allChannelTabIndex = 2;
        }

        LinkedList<String> groupList = new LinkedList<>();
        if (mPlaylist != null) {
            for (int i = 0; i < mPlaylist.getPlaylistItems().size(); i++) {
                M3UItem m3UItem = mPlaylist.getPlaylistItems().get(i);
                if (groupList.contains(m3UItem.getItemGroup())) {
                    continue;
                }
                groupList.add(m3UItem.getItemGroup());
            }
            adapter.addFragment(getString(R.string.all_channels));
            for (int i = 0; i < groupList.size(); i++) {
                String groupTitle = groupList.get(i);
                if (!TextUtils.isEmpty(groupTitle)) {
                    adapter.addFragment(groupTitle);
                }
            }
        }

        viewPager.setAdapter(adapter);
        tabLayout.setVisibility(View.VISIBLE);
        placeholderContainer.setVisibility(View.GONE);
        progressBar.setVisibility(View.GONE);
    }

    private void loadServer(String url) {
        //mProgressBar.setVisibility(View.VISIBLE);
        new LoadServer().execute(url);
    }

    @Override
    public void onBackPressed() {
        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START);
        } else {
            super.onBackPressed();
        }
    }

    @SuppressWarnings("StatementWithEmptyBody")
    @Override
    public boolean onNavigationItemSelected(MenuItem item) {
        // Handle navigation view item clicks here.
        int id = item.getItemId();

        switch (id) {
//            case R.id.nav_playlist:
//                navigateToPlaylist();
//                break;
//            case R.id.nav_live_stream:
//                openPlayStreamDialog();
//                break;
            case R.id.nav_share:
                shareApp();
                break;
            case R.id.nav_rate:
                launchMarket();
                break;
            case R.id.nav_about:
                showAbout();
                break;
            default:
                break;
        }

        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        drawer.closeDrawer(GravityCompat.START);
        return true;
    }

    private void navigateToPlaylist() {
        Router.navigateTo(this, Router.Screens.PLAYLIST);
    }

    private void requestStoragePermission() {
        if (IseaSoft.checkPermission(Manifest.permission.READ_EXTERNAL_STORAGE) && IseaSoft.checkPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
            loadChannels();
        } else {
            if (IseaSoft.shouldShowRequestPermissionRationale(this, Manifest.permission.READ_EXTERNAL_STORAGE)) {
                Snackbar.make(panelLayout, getString(R.string.request_storage_permission_message_load,
                        getString(R.string.app_name)),
                        Snackbar.LENGTH_INDEFINITE)
                        .setAction("OK", new View.OnClickListener() {
                            @Override
                            public void onClick(View view) {
                                IseaSoft.askForPermission(MainActivity.this, new String[]{Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.WRITE_EXTERNAL_STORAGE}, permissionReadstorageCallback);
                            }
                        }).show();
            } else {
                IseaSoft.askForPermission(this, new String[]{Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.WRITE_EXTERNAL_STORAGE}, permissionReadstorageCallback);
            }
        }
    }

    private void parsePlaylist(InputStream inputStream) {

        M3UParser m3UParser = new M3UParser();
        try {
            mPlaylist = m3UParser.parseFile(inputStream);
            updateUI();
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
    }

    private void updateUI() {
        new Handler(Looper.getMainLooper()).post(() -> {
            if (viewPager != null) {
                setupViewPager(viewPager);
                viewPager.setCurrentItem(allChannelTabIndex, true);//Set All channels tab
            }
        });
    }

    @Override
    public void onRequestPermissionsResult(
            int requestCode, String[] permissions, int[] grantResults) {
        IseaSoft.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    static class GroupChannelAdapter extends FragmentStatePagerAdapter {
        private final List<String> mFragmentTitles = new ArrayList<>();

        public GroupChannelAdapter(FragmentManager fm) {
            super(fm);
        }

        public void addFragment(String title) {
            mFragmentTitles.add(title);
        }

        @Override
        public Fragment getItem(int position) {
            return ChannelFragment.newInstance(mFragmentTitles.get(position));
        }

        @Override
        public int getCount() {
            return mFragmentTitles.size();
        }

        @Override
        public CharSequence getPageTitle(int position) {
            return mFragmentTitles.get(position);
        }
    }

    @SuppressLint("StaticFieldLeak")
    private class LoadServer extends AsyncTask<String, Void, Void> {

        @Override
        protected Void doInBackground(String... urls) {

            HttpHandler hh = new HttpHandler();
            InputStream inputStream = hh.makeServiceCall(urls[0]);

            parsePlaylist(inputStream);
            return null;
        }

        @Override
        protected void onPostExecute(Void aVoid) {
            super.onPostExecute(aVoid);
        }
    }

}

File: app/src/main/java/com/iseasoft/iseaiptv/ui/activity/MainActivity.kt
package com.iseasoft.iseaiptv.ui.activity

import android.Manifest
import android.annotation.SuppressLint
import android.content.Intent
import android.os.AsyncTask
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.text.TextUtils
import android.view.MenuItem
import android.view.View
import android.widget.Button
import android.widget.ProgressBar
import android.widget.TextView
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.widget.Toolbar
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentStatePagerAdapter
import androidx.fragment.app.FragmentTransaction
import androidx.viewpager.widget.ViewPager
import com.google.android.material.navigation.NavigationView
import com.google.android.material.snackbar.Snackbar
import com.google.android.material.tabs.TabLayout
import com.iseasoft.iseaiptv.App
import com.iseasoft.iseaiptv.Constants
import com.iseasoft.iseaiptv.R
import com.iseasoft.iseaiptv.helpers.Router
import com.iseasoft.iseaiptv.http.HttpHandler
import com.iseasoft.iseaiptv.models.M3UPlaylist
import com.iseasoft.iseaiptv.models.Playlist
import com.iseasoft.iseaiptv.parsers.M3UParser
import com.iseasoft.iseaiptv.permissions.IseaSoft
import com.iseasoft.iseaiptv.permissions.PermissionCallback
import com.iseasoft.iseaiptv.ui.fragment.ChannelFragment
import com.iseasoft.iseaiptv.ui.fragment.HomeFragment
import com.iseasoft.iseaiptv.utils.PreferencesUtility
import com.startapp.sdk.adsbase.Ad
import com.startapp.sdk.adsbase.StartAppAd
import com.startapp.sdk.adsbase.adlisteners.AdEventListener
import java.io.File
import java.io.FileInputStream
import java.io.FileNotFoundException
import java.io.InputStream
import java.util.*

class MainActivity : BaseActivity(), NavigationView.OnNavigationItemSelectedListener {

    private var allChannelTabIndex = 1
    private var panelLayout: CoordinatorLayout? = null
    private var viewPager: ViewPager? = null
    private var tabLayout: TabLayout? = null
    private var placeholderContainer: ConstraintLayout? = null
    private var progressBar: ProgressBar? = null

    var playlist: M3UPlaylist? = null
        private set

    private val permissionReadstorageCallback = object : PermissionCallback {
        override fun permissionGranted() {
            loadChannels()
        }

        override fun permissionRefused() {
            requestStoragePermission()
        }
    }
    private var adapter: GroupChannelAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        initStartAppSdk()
        setContentView(R.layout.activity_main)
        super.onCreate(savedInstanceState)

        val toolbar = findViewById<View>(R.id.toolbar) as Toolbar
        setSupportActionBar(toolbar)

        val drawer = findViewById<View>(R.id.drawer_layout) as DrawerLayout
        val toggle = ActionBarDrawerToggle(
                this, drawer, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close)
        drawer.addDrawerListener(toggle)
        toggle.syncState()

        val navigationView = findViewById<View>(R.id.nav_view) as NavigationView
        navigationView.setNavigationItemSelectedListener(this)

        panelLayout = findViewById(R.id.panel_layout)
        viewPager = findViewById<View>(R.id.viewpager) as ViewPager
        tabLayout = findViewById<View>(R.id.tabs) as TabLayout
        tabLayout!!.setupWithViewPager(viewPager)
        placeholderContainer = findViewById(R.id.placeholder_container)
        progressBar = findViewById(R.id.progressBar)

        checkToPlayFromPushNotification()
        loadChannels()
        setupStartAppBanner()
    }

    private fun checkToPlayFromPushNotification() {
        val intent = intent
        if (intent != null) {
            val url = intent.getStringExtra(Constants.PUSH_URL_KEY)
            val message = intent.getStringExtra(Constants.PUSH_MESSAGE)

            if (!TextUtils.isEmpty(url)) {
                if (checkPlayableUrl(url)) {
                    val playerIntent = Intent(this, PlayerActivity::class.java)
                    playerIntent.putExtra(Constants.PUSH_URL_KEY, url)
                    playerIntent.putExtra(Constants.PUSH_MESSAGE, message)
                    startActivity(playerIntent)
                } else {
                    val playlist = Playlist()
                    playlist.name = message
                    playlist.link = url
                    PreferencesUtility.getInstance(this).savePlaylist(playlist)
                }
            }
        }
    }

    private fun checkPlayableUrl(url: String): Boolean {
        return true
    }

    override fun onResume() {
        super.onResume()
    }

    private fun loadChannels() {
        val lastPlaylist = PreferencesUtility.getInstance(this).lastPlaylist
        if (lastPlaylist != null) {
            displayPlaylistInfo(lastPlaylist)
            if (lastPlaylist.link!!.trim { it <= ' ' }.startsWith("http")) {
                loadServer(lastPlaylist.link)
            } else {
                try {
                    val file = File(lastPlaylist.link)
                    val inputStream = FileInputStream(file)
                    parsePlaylist(inputStream)
                } catch (e: FileNotFoundException) {
                    e.printStackTrace()
                    updateUI()
                }

            }
        } else {
            updateUI()
        }

    }

    private fun displayPlaylistInfo(lastPlaylist: Playlist) {
        val navigationView = findViewById<View>(R.id.nav_view) as NavigationView
        val header = navigationView.getHeaderView(0)
        val playlistName = header.findViewById<TextView>(R.id.nav_header_title)
        val playlistLink = header.findViewById<TextView>(R.id.nav_header_description)
        playlistName.text = lastPlaylist.name
        //playlistLink.setText(lastPlaylist.getLink());
        supportActionBar!!.title = lastPlaylist.name
    }

    private fun showChannelPlaceholder() {
        tabLayout!!.visibility = View.GONE
        progressBar!!.visibility = View.GONE
        placeholderContainer!!.visibility = View.VISIBLE
        val btnAdd = findViewById<Button>(R.id.btn_add_playlist)
        btnAdd.setOnClickListener { navigateToPlaylist() }
    }

    private fun setupViewPager(viewPager: ViewPager) {
        if (adapter == null) {
            adapter = GroupChannelAdapter(supportFragmentManager)
        }
        adapter!!.addFragment(getString(R.string.favorites))
        if (!PreferencesUtility.getInstance(this).hasNoHistoryWatching()) {
            adapter!!.addFragment(getString(R.string.history_watching))
            allChannelTabIndex = 2
        }

        val groupList = LinkedList<String>()
        if (playlist != null) {
            for (i in 0 until playlist!!.playlistItems!!.size) {
                val m3UItem = playlist!!.playlistItems!![i]
                if (groupList.contains(m3UItem.itemGroup)) {
                    continue
                }
                m3UItem.itemGroup?.let { groupList.add(it) }
            }
            adapter!!.addFragment(getString(R.string.all_channels))
            for (i in groupList.indices) {
                val groupTitle = groupList[i]
                if (!TextUtils.isEmpty(groupTitle)) {
                    adapter!!.addFragment(groupTitle)
                }
            }
        }

        viewPager.adapter = adapter
        tabLayout!!.visibility = View.VISIBLE
        placeholderContainer!!.visibility = View.GONE
        progressBar!!.visibility = View.GONE
    }

    private fun loadServer(url: String?) {
        //mProgressBar.setVisibility(View.VISIBLE);
        LoadServer().execute(url)
    }

    override fun onBackPressed() {
        val drawer = findViewById<View>(R.id.drawer_layout) as DrawerLayout
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START)
        } else {
            super.onBackPressed()
        }
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        // Handle navigation view item clicks here.
        val id = item.itemId

        when (id) {
            //            case R.id.nav_playlist:
            //                navigateToPlaylist();
            //                break;
            R.id.nav_trending_app -> openTrendingApps()
            R.id.nav_share -> shareApp()
            R.id.nav_rate -> launchMarket()
            R.id.nav_about -> showAbout()
            else -> {
            }
        }

        val drawer = findViewById<View>(R.id.drawer_layout) as DrawerLayout
        drawer.closeDrawer(GravityCompat.START)
        return true
    }

    private fun openTrendingApps() {
        val offerWallAds = StartAppAd(this)
        offerWallAds.loadAd(StartAppAd.AdMode.OFFERWALL, object : AdEventListener {
            override fun onReceiveAd(ad: Ad) {
                offerWallAds.showAd()
            }

            override fun onFailedToReceiveAd(ad: Ad) {
                openTrendingApps()
            }
        })
    }

    private fun navigateToPlaylist() {
        Router.navigateTo(this, Router.Screens.PLAYLIST)
    }

    @SuppressLint("StringFormatInvalid")
    private fun requestStoragePermission() {
        if (IseaSoft.checkPermission(Manifest.permission.READ_EXTERNAL_STORAGE) && IseaSoft.checkPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
            loadChannels()
        } else {
            if (IseaSoft.shouldShowRequestPermissionRationale(this, Manifest.permission.READ_EXTERNAL_STORAGE)) {
                Snackbar.make(panelLayout!!, getString(R.string.request_storage_permission_message_load,
                        getString(R.string.app_name)),
                        Snackbar.LENGTH_INDEFINITE)
                        .setAction("OK") { IseaSoft.askForPermission(this@MainActivity, arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.WRITE_EXTERNAL_STORAGE), permissionReadstorageCallback) }.show()
            } else {
                IseaSoft.askForPermission(this, arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.WRITE_EXTERNAL_STORAGE), permissionReadstorageCallback)
            }
        }
    }

    private fun parsePlaylist(inputStream: InputStream?) {

        val m3UParser = M3UParser()
        try {
            playlist = inputStream?.let { m3UParser.parseFile(it) }
            playlist?.playlistItems?.let {
                App.channelList = playlist?.playlistItems!!
            }
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        }

    }

    private fun updateUI() {
        //            if (viewPager != null) {
        //                setupViewPager(viewPager);
        //                viewPager.setCurrentItem(allChannelTabIndex, true);//Set All channels tab
        //            }
        Handler(Looper.getMainLooper()).post { this.setupHomeView() }
    }

    private fun setupHomeView() {
        val fm = supportFragmentManager
        val ft = fm.beginTransaction()

        val homeFragment = HomeFragment()
        ft.replace(R.id.home_content, homeFragment, HomeFragment.TAG)
        ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
        ft.commitAllowingStateLoss()
        placeholderContainer!!.visibility = View.GONE
        progressBar!!.visibility = View.GONE
    }

    override fun onRequestPermissionsResult(
            requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        IseaSoft.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    internal class GroupChannelAdapter(fm: FragmentManager) : FragmentStatePagerAdapter(fm) {
        private val mFragmentTitles = ArrayList<String>()

        fun addFragment(title: String) {
            mFragmentTitles.add(title)
        }

        override fun getItem(position: Int): Fragment {
            return ChannelFragment.newInstance(mFragmentTitles[position])
        }

        override fun getCount(): Int {
            return mFragmentTitles.size
        }

        override fun getPageTitle(position: Int): CharSequence? {
            return mFragmentTitles[position]
        }
    }

    @SuppressLint("StaticFieldLeak")
    private inner class LoadServer : AsyncTask<String, Void, Void?>() {

        override fun doInBackground(vararg urls: String): Void? {

            val hh = HttpHandler()
            val inputStream = hh.makeServiceCall(urls[0])

            parsePlaylist(inputStream)
            return null
        }

        override fun onPostExecute(aVoid: Void?) {
            super.onPostExecute(aVoid)
            updateUI()
        }
    }

}
